Entities/Meta/CTFMusic.as

```diff
@@ -0,0 +1,289 @@
+const string[] blobList = {"knight", "archer", "builder", "ballista", "tunnel", "keg"};
+
+enum GameMusicTag
+{
+	world_ambient,
+
+	world_music_start,
+
+	world_ambient_underground,
+	world_ambient_mountain,
+	world_intro,
+	world_home,
+	world_calm,
+	world_battle,
+	world_outro,
+
+	world_music_end,
+};
+
+string base_blob_name = "hall";
+
+void onInit(CBlob@ this)
+{
+	CMixer@ mixer = getMixer();
+	if (mixer is null)
+		return;
+
+	mixer.ResetMixer();
+	this.set_bool("initialized game", false);
+
+	CRules@ rules = getRules();
+	if (rules !is null)
+	{
+		if (rules.exists("music - base name"))
+		{
+			base_blob_name = rules.get_string("music - base name");
+		}
+	}
+}
+
+void onTick(CBlob@ this)
+{
+	CMixer@ mixer = getMixer();
+	if (mixer is null)
+		return;
+
+	if (s_soundon != 0 && s_musicvolume > 0.0f)
+	{
+		if (!this.get_bool("initialized game"))
+		{
+			AddGameMusic(this, mixer);
+		}
+
+		GameMusicLogic(this, mixer);
+	}
+	else
+	{
+		mixer.FadeOutAll(0.0f, 2.0f);
+	}
+}
+
+//sound references with tag
+void AddGameMusic(CBlob@ this, CMixer@ mixer)
+{
+	if (mixer is null)
+		return;
+
+	this.set_bool("initialized game", true);
+	mixer.ResetMixer();
+	mixer.AddTrack("Sounds/Music/ambient_forest.ogg", world_ambient);
+	mixer.AddTrack("Sounds/Music/ambient_mountain.ogg", world_ambient_mountain);
+	mixer.AddTrack("Sounds/Music/ambient_cavern.ogg", world_ambient_underground);
+	mixer.AddTrack("Sounds/Music/KAGWorldIntroShortA.ogg", world_intro);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-1a.ogg", world_calm);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-2a.ogg", world_calm);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-3a.ogg", world_calm);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-4a.ogg", world_calm);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-5a.ogg", world_home);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-6a.ogg", world_home);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-7a.ogg", world_home);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-8a.ogg", world_home);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-9a.ogg", world_calm);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-10a.ogg", world_battle);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-11a.ogg", world_battle);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-12a.ogg", world_battle);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-13.ogg", world_battle);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-14.ogg", world_battle);
+	mixer.AddTrack("Sounds/Music/KAGWorld1-14outro.ogg", world_outro);
+}
+
+uint timer = 0;
+bool gameStarted = true;
+
+void GameMusicLogic(CBlob@ this, CMixer@ mixer)
+{
+	timer++;
+	if (mixer is null || !s_gamemusic)
+		return;
+
+	CRules @rules = getRules();
+	CBlob @blob = getLocalPlayerBlob();
+	if (blob is null)
+	{
+		mixer.FadeOutAll(0.0f, 6.0f);
+		return;
+	}
+
+	CMap@ map = blob.getMap();
+	if (map is null)
+		return;
+
+	if (rules.isWarmup())
+	{
+		gameStarted = false;
+		
+		if (timer % getTicksASecond() * 3 == 0)
+		{
+			Vec2f pos = blob.getPosition();
```

Comment:
blob.getPosition and checkUnderground can be pulled out to line ~116 because it's done in both major branches.